File: ios/MyNasWidgets/Shared/WidgetDataManager.swift
//
//  WidgetDataManager.swift
//  MyNasWidgets
//
//  Shared data manager for all widgets
//

import Foundation
import SwiftUI

/// 小组件数据管理器
/// 负责从 App Group 共享存储中读取数据
class WidgetDataManager {
    static let shared = WidgetDataManager()

    private let appGroupId = "group.com.kkape.mynas"
    private var userDefaults: UserDefaults? {
        UserDefaults(suiteName: appGroupId)
    }

    // MARK: - Storage Data

    struct StorageData: Codable {
        let totalBytes: Int
        let usedBytes: Int
        let nasName: String
        let adapterType: String
        let lastUpdated: Int
        let isConnected: Bool

        var usagePercent: Double {
            totalBytes > 0 ? Double(usedBytes) / Double(totalBytes) : 0
        }

        var isLowSpace: Bool {
            usagePercent > 0.9
        }

        var hasValidData: Bool {
            totalBytes > 0
        }

        static let placeholder = StorageData(
            totalBytes: 1_000_000_000_000,
            usedBytes: 650_000_000_000,
            nasName: "My NAS",
            adapterType: "synology",
            lastUpdated: Int(Date().timeIntervalSince1970 * 1000),
            isConnected: true
        )

        static let empty = StorageData(
            totalBytes: 0,
            usedBytes: 0,
            nasName: "",
            adapterType: "unknown",
            lastUpdated: 0,
            isConnected: false
        )
    }

    func getStorageData() -> StorageData {
        guard let defaults = userDefaults,
              let jsonData = defaults.data(forKey: "widget_storage_data"),
              let data = try? JSONDecoder().decode(StorageData.self, from: jsonData) else {
            return .empty
        }
        return data
    }

    // MARK: - Download Data

    struct DownloadTaskSummary: Codable {
        let id: String
        let fileName: String
        let progress: Double
        let status: String
    }

    struct DownloadData: Codable {
        let activeTasks: [DownloadTaskSummary]
        let completedCount: Int
        let totalCount: Int
        let lastUpdated: Int

        var activeCount: Int {
            activeTasks.count
        }

        var hasActiveDownloads: Bool {
            !activeTasks.isEmpty
        }

        var overallProgress: Double {
            guard !activeTasks.isEmpty else { return 0 }
            let sum = activeTasks.reduce(0.0) { $0 + $1.progress }
            return sum / Double(activeTasks.count)
        }

        var currentFileName: String? {
            activeTasks.first?.fileName
        }

        static let placeholder = DownloadData(
            activeTasks: [
                DownloadTaskSummary(id: "1", fileName: "movie.mkv", progress: 0.45, status: "downloading")
            ],
            completedCount: 3,
            totalCount: 5,
            lastUpdated: Int(Date().timeIntervalSince1970 * 1000)
        )

        static let empty = DownloadData(
            activeTasks: [],
            completedCount: 0,
            totalCount: 0,
            lastUpdated: 0
        )
    }

    func getDownloadData() -> DownloadData {
        guard let defaults = userDefaults,
              let jsonData = defaults.data(forKey: "widget_download_data"),
              let data = try? JSONDecoder().decode(DownloadData.self, from: jsonData) else {
            return .empty
        }
        return data
    }

    // MARK: - Media Data

    struct MediaData: Codable {
        let title: String?
        let artist: String?
        let album: String?
        let coverImagePath: String?
        let isPlaying: Bool
        let progress: Double
        let currentTime: Int
        let totalTime: Int
        let themeColor: Int?

        var hasContent: Bool {
            title != nil && !(title?.isEmpty ?? true)
        }

        static let placeholder = MediaData(
            title: "Song Title",
            artist: "Artist Name",
            album: "Album Name",
            coverImagePath: nil,
            isPlaying: true,
            progress: 0.45,
            currentTime: 120,
            totalTime: 300,
            themeColor: nil
        )

        static let empty = MediaData(
            title: nil,
            artist: nil,
            album: nil,
            coverImagePath: nil,
            isPlaying: false,
            progress: 0,
            currentTime: 0,
            totalTime: 0,
            themeColor: nil
        )
    }

    func getMediaData() -> MediaData {
        guard let defaults = userDefaults,
              let jsonData = defaults.data(forKey: "widget_media_data"),
              let data = try? JSONDecoder().decode(MediaData.self, from: jsonData) else {
            return .empty
        }
        return data
    }

    /// 获取封面图片
    func getCoverImage() -> UIImage? {
        guard let defaults = userDefaults,
              let imageData = defaults.data(forKey: "widget_cover_image") else {
            return nil
        }
        return UIImage(data: imageData)
    }

    // MARK: - Quick Access Data

    struct QuickAccessData: Codable {
        let items: [String]
        let nasName: String?
        let isConnected: Bool

        static let defaultData = QuickAccessData(
            items: ["music", "video", "book"],
            nasName: nil,
            isConnected: false
        )
    }

    func getQuickAccessData() -> QuickAccessData {
        guard let defaults = userDefaults,
              let jsonData = defaults.data(forKey: "widget_quick_access_data"),
              let data = try? JSONDecoder().decode(QuickAccessData.self, from: jsonData) else {
            return .defaultData
        }
        return data
    }

    // MARK: - Connection Status

    func isNasConnected() -> Bool {
        userDefaults?.bool(forKey: "widget_nas_connected") ?? false
    }

    func getNasName() -> String? {
        userDefaults?.string(forKey: "widget_nas_name")
    }
}

// MARK: - Utility Extensions

extension Int {
    /// 格式化字节数为人类可读的字符串
    var formattedBytes: String {
        let bytes = Double(self)
        let kb = bytes / 1024
        let mb = kb / 1024
        let gb = mb / 1024
        let tb = gb / 1024

        if tb >= 1 {
            return String(format: "%.1f TB", tb)
        } else if gb >= 1 {
            return String(format: "%.1f GB", gb)
        } else if mb >= 1 {
            return String(format: "%.1f MB", mb)
        } else if kb >= 1 {
            return String(format: "%.1f KB", kb)
        } else {
            return "\(self) B"
        }
    }
}

extension Int {
    /// 格式化秒数为时间字符串
    var formattedDuration: String {
        let hours = self / 3600
        let minutes = (self % 3600) / 60
        let seconds = self % 60

        if hours > 0 {
            return String(format: "%d:%02d:%02d", hours, minutes, seconds)
        } else {
            return String(format: "%d:%02d", minutes, seconds)
        }
    }
}

extension Color {
    /// 从 ARGB 整数创建颜色
    init(argb: Int) {
        let a = Double((argb >> 24) & 0xFF) / 255.0
        let r = Double((argb >> 16) & 0xFF) / 255.0
        let g = Double((argb >> 8) & 0xFF) / 255.0
        let b = Double(argb & 0xFF) / 255.0
        self.init(.sRGB, red: r, green: g, blue: b, opacity: a)
    }

    /// 从十六进制字符串创建颜色
    init(hex: String) {
        let hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&int)
        let r, g, b: Double
        switch hex.count {
        case 6:
            r = Double((int >> 16) & 0xFF) / 255.0
            g = Double((int >> 8) & 0xFF) / 255.0
            b = Double(int & 0xFF) / 255.0
        default:
            r = 1
            g = 1
            b = 1
        }
        self.init(.sRGB, red: r, green: g, blue: b, opacity: 1)
    }
}

File: android/app/src/main/kotlin/com/kkape/mynas/widgets/WidgetDataManager.kt
package com.kkape.mynas.widgets

import android.content.Context
import android.content.SharedPreferences
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Widget 数据管理器
 * 负责读写共享存储中的 Widget 数据
 */
class WidgetDataManager(private val context: Context) {

    private val prefs: SharedPreferences = context.getSharedPreferences(
        PREFS_NAME, Context.MODE_PRIVATE
    )

    companion object {
        const val PREFS_NAME = "widget_data"

        // Storage Widget keys
        const val KEY_STORAGE_DATA = "storage_data"
        const val KEY_NAS_CONNECTED = "nas_connected"
        const val KEY_NAS_NAME = "nas_name"

        // Download Widget keys
        const val KEY_DOWNLOAD_DATA = "download_data"

        // Quick Access Widget keys
        const val KEY_QUICK_ACCESS_DATA = "quick_access_data"

        // Media Widget keys
        const val KEY_MEDIA_DATA = "media_data"
        const val KEY_COVER_IMAGE_PATH = "cover_image_path"
    }

    // ==================== Storage Data ====================

    data class StorageData(
        val totalBytes: Long,
        val usedBytes: Long,
        val nasName: String,
        val adapterType: String,
        val lastUpdated: Long,
        val isConnected: Boolean
    ) {
        val usagePercent: Double
            get() = if (totalBytes > 0) usedBytes.toDouble() / totalBytes else 0.0

        val usagePercentInt: Int
            get() = (usagePercent * 100).toInt()

        val isLowSpace: Boolean
            get() = usagePercent > 0.9

        val hasValidData: Boolean
            get() = totalBytes > 0

        companion object {
            val EMPTY = StorageData(0, 0, "", "unknown", 0, false)

            val PLACEHOLDER = StorageData(
                1_000_000_000_000L,
                650_000_000_000L,
                "My NAS",
                "synology",
                System.currentTimeMillis(),
                true
            )

            fun fromJson(json: JSONObject): StorageData {
                return StorageData(
                    totalBytes = json.optLong("totalBytes", 0),
                    usedBytes = json.optLong("usedBytes", 0),
                    nasName = json.optString("nasName", ""),
                    adapterType = json.optString("adapterType", "unknown"),
                    lastUpdated = json.optLong("lastUpdated", 0),
                    isConnected = json.optBoolean("isConnected", false)
                )
            }
        }
    }

    fun getStorageData(): StorageData {
        val jsonStr = prefs.getString(KEY_STORAGE_DATA, null) ?: return StorageData.EMPTY
        return try {
            StorageData.fromJson(JSONObject(jsonStr))
        } catch (e: Exception) {
            StorageData.EMPTY
        }
    }

    fun saveStorageData(data: Map<String, Any?>) {
        val json = JSONObject(data)
        prefs.edit()
            .putString(KEY_STORAGE_DATA, json.toString())
            .putBoolean(KEY_NAS_CONNECTED, data["isConnected"] as? Boolean ?: false)
            .putString(KEY_NAS_NAME, data["nasName"] as? String)
            .apply()
    }

    // ==================== Download Data ====================

    data class DownloadTaskSummary(
        val id: String,
        val fileName: String,
        val progress: Double,
        val status: String
    ) {
        companion object {
            fun fromJson(json: JSONObject): DownloadTaskSummary {
                return DownloadTaskSummary(
                    id = json.optString("id", ""),
                    fileName = json.optString("fileName", ""),
                    progress = json.optDouble("progress", 0.0),
                    status = json.optString("status", "pending")
                )
            }
        }
    }

    data class DownloadData(
        val activeTasks: List<DownloadTaskSummary>,
        val completedCount: Int,
        val totalCount: Int,
        val lastUpdated: Long
    ) {
        val activeCount: Int
            get() = activeTasks.size

        val hasActiveDownloads: Boolean
            get() = activeTasks.isNotEmpty()

        val overallProgress: Double
            get() = if (activeTasks.isEmpty()) 0.0
            else activeTasks.sumOf { it.progress } / activeTasks.size

        val currentFileName: String?
            get() = activeTasks.firstOrNull()?.fileName

        companion object {
            val EMPTY = DownloadData(emptyList(), 0, 0, 0)

            val PLACEHOLDER = DownloadData(
                listOf(DownloadTaskSummary("1", "movie.mkv", 0.45, "downloading")),
                3,
                5,
                System.currentTimeMillis()
            )

            fun fromJson(json: JSONObject): DownloadData {
                val tasksArray = json.optJSONArray("activeTasks") ?: JSONArray()
                val tasks = (0 until tasksArray.length()).map {
                    DownloadTaskSummary.fromJson(tasksArray.getJSONObject(it))
                }
                return DownloadData(
                    activeTasks = tasks,
                    completedCount = json.optInt("completedCount", 0),
                    totalCount = json.optInt("totalCount", 0),
                    lastUpdated = json.optLong("lastUpdated", 0)
                )
            }
        }
    }

    fun getDownloadData(): DownloadData {
        val jsonStr = prefs.getString(KEY_DOWNLOAD_DATA, null) ?: return DownloadData.EMPTY
        return try {
            DownloadData.fromJson(JSONObject(jsonStr))
        } catch (e: Exception) {
            DownloadData.EMPTY
        }
    }

    fun saveDownloadData(data: Map<String, Any?>) {
        val json = JSONObject(data)
        prefs.edit()
            .putString(KEY_DOWNLOAD_DATA, json.toString())
            .apply()
    }

    // ==================== Quick Access Data ====================

    data class QuickAccessData(
        val items: List<String>,
        val nasName: String?,
        val isConnected: Boolean
    ) {
        companion object {
            // 默认快捷操作项 - 使用与路由一致的名称
            val DEFAULT = QuickAccessData(listOf("music", "video", "reading"), null, false)

            fun fromJson(json: JSONObject): QuickAccessData {
                val itemsArray = json.optJSONArray("items") ?: JSONArray()
                val items = (0 until itemsArray.length()).map { itemsArray.getString(it) }
                return QuickAccessData(
                    items = items.ifEmpty { listOf("music", "video", "reading") },
                    nasName = json.optString("nasName", null),
                    isConnected = json.optBoolean("isConnected", false)
                )
            }
        }
    }

    fun getQuickAccessData(): QuickAccessData {
        val jsonStr = prefs.getString(KEY_QUICK_ACCESS_DATA, null)
            ?: return QuickAccessData.DEFAULT
        return try {
            QuickAccessData.fromJson(JSONObject(jsonStr))
        } catch (e: Exception) {
            QuickAccessData.DEFAULT
        }
    }

    fun saveQuickAccessData(data: Map<String, Any?>) {
        val json = JSONObject(data)
        prefs.edit()
            .putString(KEY_QUICK_ACCESS_DATA, json.toString())
            .apply()
    }

    // ==================== Media Data ====================

    data class MediaData(
        val title: String?,
        val artist: String?,
        val album: String?,
        val coverImagePath: String?,
        val isPlaying: Boolean,
        val progress: Double,
        val currentTime: Int,
        val totalTime: Int,
        val themeColor: Int?
    ) {
        val hasContent: Boolean
            get() = !title.isNullOrEmpty()

        companion object {
            val EMPTY = MediaData(null, null, null, null, false, 0.0, 0, 0, null)

            val PLACEHOLDER = MediaData(
                "Song Title",
                "Artist Name",
                "Album Name",
                null,
                true,
                0.45,
                120,
                300,
                null
            )

            fun fromJson(json: JSONObject): MediaData {
                return MediaData(
                    title = json.optString("title", null),
                    artist = json.optString("artist", null),
                    album = json.optString("album", null),
                    coverImagePath = json.optString("coverImagePath", null),
                    isPlaying = json.optBoolean("isPlaying", false),
                    progress = json.optDouble("progress", 0.0),
                    currentTime = json.optInt("currentTime", 0),
                    totalTime = json.optInt("totalTime", 0),
                    themeColor = if (json.has("themeColor")) json.optInt("themeColor") else null
                )
            }
        }
    }

    fun getMediaData(): MediaData {
        val jsonStr = prefs.getString(KEY_MEDIA_DATA, null) ?: return MediaData.EMPTY
        return try {
            MediaData.fromJson(JSONObject(jsonStr))
        } catch (e: Exception) {
            MediaData.EMPTY
        }
    }

    fun saveMediaData(data: Map<String, Any?>) {
        // Remove cover image data before saving to JSON
        val jsonData = data.toMutableMap()
        jsonData.remove("coverImageData")

        val json = JSONObject(jsonData)
        prefs.edit()
            .putString(KEY_MEDIA_DATA, json.toString())
            .apply()
    }

    fun saveCoverImage(imageData: ByteArray) {
        val file = File(context.cacheDir, "widget_cover.jpg")
        file.writeBytes(imageData)
        prefs.edit()
            .putString(KEY_COVER_IMAGE_PATH, file.absolutePath)
            .apply()
    }

    fun getCoverImage(): Bitmap? {
        val path = prefs.getString(KEY_COVER_IMAGE_PATH, null) ?: return null
        val file = File(path)
        if (!file.exists()) return null
        return try {
            BitmapFactory.decodeFile(path)
        } catch (e: Exception) {
            null
        }
    }

    // ==================== Utility ====================

    fun isNasConnected(): Boolean {
        return prefs.getBoolean(KEY_NAS_CONNECTED, false)
    }

    fun getNasName(): String? {
        return prefs.getString(KEY_NAS_NAME, null)
    }
}

/**
 * 格式化字节数为人类可读的字符串
 */
fun Long.formatBytes(): String {
    val bytes = this.toDouble()
    val kb = bytes / 1024
    val mb = kb / 1024
    val gb = mb / 1024
    val tb = gb / 1024

    return when {
        tb >= 1 -> String.format("%.1f TB", tb)
        gb >= 1 -> String.format("%.1f GB", gb)
        mb >= 1 -> String.format("%.1f MB", mb)
        kb >= 1 -> String.format("%.1f KB", kb)
        else -> "$this B"
    }
}

/**
 * 格式化秒数为时间字符串
 */
fun Int.formatDuration(): String {
    val hours = this / 3600
    val minutes = (this % 3600) / 60
    val seconds = this % 60

    return if (hours > 0) {
        String.format("%d:%02d:%02d", hours, minutes, seconds)
    } else {
        String.format("%d:%02d", minutes, seconds)
    }
}
